src/clp_ffi_py/ir/native/PyDeserializer.cpp

```diff
@@ -0,0 +1,281 @@
+
+        PyDeserializer::IrUnitHandler::SchemaTreeNodeInsertionHandle
+                trivial_schema_tree_node_insertion_handle
+                = []([[maybe_unused]] clp::ffi::SchemaTree::NodeLocator) -> IRErrorCode {
+            return IRErrorCode::IRErrorCode_Success;
+        };
+
+        PyDeserializer::IrUnitHandler::EndOfStreamHandle end_of_stream_handle
+                = [this]() -> IRErrorCode { return this->handle_end_of_stream(); };
+
+        auto deserializer_result{Deserializer::create(
+                *m_deserializer_buffer_reader,
+                {std::move(log_event_handle),
+                 std::move(trivial_utc_offset_handle),
+                 std::move(trivial_schema_tree_node_insertion_handle),
+                 std::move(end_of_stream_handle)}
+        )};
+        if (deserializer_result.has_error()) {
+            PyErr_Format(
+                    PyExc_RuntimeError,
+                    cDeserializerCreateErrorFormatStr.data(),
+                    deserializer_result.error().message().c_str()
+            );
+            return false;
+        }
+        m_deserializer = new clp::ffi::ir_stream::Deserializer<PyDeserializer::IrUnitHandler>{
+                std::move(deserializer_result.value())
+        };
+    } catch (clp::TraceableException& exception) {
+        handle_traceable_exception(exception);
+        return false;
+    }
+
+    return true;
+}
+
+auto PyDeserializer::deserialize_log_event() -> PyObject* {
+    try {
+        while (false == is_stream_complete()) {
+            auto const ir_unit_type_result{
+                    m_deserializer->deserialize_next_ir_unit(*m_deserializer_buffer_reader)
+            };
+            if (ir_unit_type_result.has_error()) {
+                if (false == handle_incomplete_ir_error(ir_unit_type_result.error())) {
+                    return nullptr;
+                }
+                break;
+            }
+            auto const ir_unit_type{ir_unit_type_result.value()};
+            if (IrUnitType::LogEvent == ir_unit_type && has_unreleased_deserialized_log_event()) {
+                return py_reinterpret_cast<PyObject>(
+                        PyKeyValuePairLogEvent::create(release_deserialized_log_event())
+                );
+            }
+        }
+    } catch (clp::TraceableException& exception) {
+        handle_traceable_exception(exception);
+        return nullptr;
+    }
+
+    Py_RETURN_NONE;
+}
+
+auto PyDeserializer::module_level_init(PyObject* py_module) -> bool {
+    static_assert(std::is_trivially_destructible<PyDeserializer>());
+    auto* type{py_reinterpret_cast<PyTypeObject>(PyType_FromSpec(&PyDeserializer_type_spec))};
+    m_py_type.reset(type);
+    if (nullptr == type) {
+        return false;
+    }
+    return add_python_type(get_py_type(), "Deserializer", py_module);
+}
+
+auto PyDeserializer::handle_log_event(clp::ffi::KeyValuePairLogEvent&& log_event) -> IRErrorCode {
+    if (has_unreleased_deserialized_log_event()) {
```

Comment:
In normal operation, if no code runs into error, do you see a possibility that when this method is called, there is unrealised deserialized_log_event?

Reply:
No. It's only possible if an error happened

Reply:
The error must already be returned and handled before control flow reaches here.

Also this is more like a sanity check. This is not like unique pointer where overwriting the pointer will automatically free the underlying object. Without releasing non-empty log event it would lead to memory leak.

Reply:
Added a comment to explain why this is required